File: fsharp/space-age/SpaceAge.fs
﻿module SpaceAge

type Planet = 
    | Mercury
    | Venus
    | Earth
    | Mars
    | Jupiter
    | Saturn
    | Neptune
    | Uranus

let private planetPeriodsInEarthYears = 
    [ Mercury, 0.2408467
      Venus,   0.61519726
      Earth,   1.0
      Mars,    1.8808158
      Jupiter, 11.862615
      Saturn,  29.447498
      Uranus,  84.016846
      Neptune, 164.79132 ] 
    |> Map.ofList

let age planet (seconds: int64) =
    let secondsOnEarth = 31557600.0
    let yearsUsingPeriod (period: float) = System.Math.Round((float seconds / period) / secondsOnEarth, 2)
        
    yearsUsingPeriod planetPeriodsInEarthYears.[planet]

File: clojure/space-age/src/space_age.clj
(ns space-age)

(def earth-years-for-planet
  { :mercury 0.2408467
    :venus 0.61519726
    :earth 1.0
    :mars 1.8808158
    :jupiter 11.862615
    :saturn 29.447498
    :uranus 84.016846
    :neptune 164.79132 })

(defn on-planet [planet seconds]
  (let [earth-years (/ seconds 31557600)]
    (* (/ 1 (get earth-years-for-planet planet)) earth-years)))

(def on-mercury (partial on-planet :mercury))
(def on-venus (partial on-planet :venus))
(def on-earth (partial on-planet :earth))
(def on-mars (partial on-planet :mars))
(def on-jupiter (partial on-planet :jupiter))
(def on-saturn (partial on-planet :saturn))
(def on-neptune (partial on-planet :neptune))
(def on-uranus (partial on-planet :uranus))

